draft-ietf-jsonpath-base.md

```diff
@@ -1355,6 +1351,7 @@ Queries:
 | `$..[0]` | `5` <br> `{"j": 4}` | `$['a'][0]` <br> `$['a'][2][0]` | Array values       |
 | `$..[0]` | `{"j": 4}` <br> `5` | `$['a'][2][0]` <br> `$['a'][0]` | Alternative result |
 | `$..[*]` <br> `$..*` | `{"j": 1, "k" : 2}` <br> `[5, 3, [{"j": 4}]]` <br> `1` <br> `2` <br> `5` <br> `3` <br> `[{"j": 4}]` <br> `{"j": 4}` <br> `4` | `$['o']` <br> `$['a']` <br> `$['o']['j']` <br> `$['o']['k']` <br> `$['a'][0]` <br> `$['a'][1]` <br> `$['a'][2]` <br> `$['a'][2][0]` <br> `$['a'][2][0]['j']` | All values    |
+| `$..o`   | `{"j": 1, "k": 2}` | `$['o']` | Input value is included in the enumeration |
```

Comment:
How does this example differ from the `..j` examples?

Reply:
> How does this example differ from the `..j` examples?

The `$..j` example is unaffected by whether or not the input value is part of the enumeration. On the other hand, this example depends on that decision. If the input value was not part of the enumeration, then the resultant nodelist would be empty. So this example demonstrates that the input value is part of the enumeration.

Reply:
I don't see how that's the case.  The `..j` examples pull the values at those locations. The `..o` example does the same thing.

Reply:
I wonder if the wording of the semantics section needs improving as it seems that you aren't reading it the same way as me.

There are two steps involved:

1. enumerating a nodelist of descendants _plus the input value_, and
2. applying the child segment corresponding to the descendant segment to that nodelist.

With that in mind, does my [previous comment](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/266#discussion_r996497750) now explain the need for the `$..o` example?

We could add a `$..*` example, but please note this would return just the descendants of the input value since step 2 would apply the child segment `$[*]` which returns all the children of each of its input values (and not the value itself).

Reply:
BTW please note that table 2 already includes a `$..*` example with the result:

> all member values and array elements contained in input value

Reply:
> With that in mind, does my previous comment now explain the need for the $..o example?

I'm sorry, I still don't understand why this example illustrates the change.  What would the result be without the change (or, what's the alternative interpretation of `..o`)?

Reply:
To be clear, I agree that the unit node should be included. I just don't understand how this example drives that point home.

Reply:
> I'm sorry, I still don't understand why this example illustrates the change. What would the result be without the change (or, what's the alternative interpretation of `..o`)?

Without the change, the enumerated nodelist would be something like this:
* `{"j": 1, "k": 2}`
* `1`
* `2`
* `[5, 3, [{"j": 4}]]`
* `5`
* `3`
* `[{"j": 4}]`
* `4`

Applying the corresponding child segment `['o']` to this list produces the empty nodelist.

With the change, the enumerated nodelist would include: `{
  "o": {"j": 1, "k": 2},
  "a": [5, 3, [{"j": 4}]]
}` and so applying the corresponding child segment `['o']` produces the nodelist:
* `{"j": 1, "k": 2}`

Reply:
Hmmm... I don't see the `..` operating separately from the `o`.  I see it all as a single step because it's defined that way in the syntax.

`..o` says, "navigate recursively and return nodes under `o` properties."  That immediately gives you 

```json
{ "j": 1, "k": 2 }
```

It doesn't say, "navigate recursively, give me all the nodes, _then from that_ give me the `o` properties."

Reply:
I now see that you're changing the text to mean the two-step operation, but again, I think that conflicts with the syntax.

I really think the simpler approach is to process the selection _while_ enumerating the (recursive) descendants.

(I know that we're not pushing algorithms, but _something_ needs to be in the spec.)

Reply:
I agree with Greg here: Just as a child segment offers `self` to the selectors, a descendants segment offers `self` and `descendant` to the selectors.  Having the `$..o` example to illustrate this is a good idea.

(This and a few other examples need to be fixed if we stick with `"j j"` and `"k.k"`.)

Reply:
Please see 17835689a97eae026254e7cc28197817cbfbcf21 in which I've expressed the semantics the way you suggest.

I believe the semantics are unchanged.
 
> (Other examples need to be fixed if we stick with `"j j"` and `"k.k"`.)

Not sure what this is referring to because the examples in this section mention neither `"j j"` nor `"k.k"`. However, I've pushed 20bb114990b615da55a7be662e5bb37e218182b7 which deletes the erroneous child selector examples.

Reply:
> However, I've pushed 20bb114 which deletes the erroneous child selector examples.

Thanks.